### PR TITLE
JWT gem

### PR DESCRIPTION
**Pull request by Charchit507**

### CODE REVIEW

File: app/controllers/concerns/json_web_token.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+require "jwt"
+module JsonWebToken
+  extend ActiveSupport::Concern
+  SECRET_KEY = Rails.application.secret_key_base
```
Comment (neeteshgupta95) [inline app/controllers/concerns/json_web_token.rb:4]: move this to `credentials.yml.enc` file and use as a variable

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  #     render json: { user: @user, token: token }, status: :ok
+  #   else
+  #     render json: { error: "Invalid username or password" }, status: :unprocessable_entity
+  #   end
+  # end
+
+  # private
+  # def user_params
+  #    params.require(:user).permit(:username, :password)
+  # end
```
Comment (neeteshgupta95) [inline app/controllers/users_controller.rb:26]: why commented code?

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+  skip_before_action :authenticate_request, only: [:create]
+  before_action :set_user, only: [:show, :destroy]
+
+  def index
+    @users = User.all
+    render json: @users, status: :ok
+  end
```
Comment (neeteshgupta95) [inline app/controllers/users_controller.rb:34]: are we listing users?

File: config/routes.rb [hunk 4]
```diff
@@ -1,6 +1,5 @@
 Rails.application.routes.draw do
-  # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
+  resource :users
```
Comment (neeteshgupta95) [inline config/routes.rb:2]: why `resource`

File: db/migrate/20220915091245_create_users.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+class CreateUsers < ActiveRecord::Migration[7.0]
+  def change
+    create_table :users do |t|
+      t.string :email
```
Comment (neeteshgupta95) [inline db/migrate/20220915091245_create_users.rb:4]: add uniqueness at the database level

File: app/models/user.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class User < ApplicationRecord
+  has_secure_password
+  validates :email, presence: true
```
Comment (neeteshgupta95) [inline app/models/user.rb:3]: add uniqueness and email regex as well

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+
+  def show
+    render json: @user, status: :ok
+  end
+
+  def create
+    @user = User.new(user_params)
```
Comment (neeteshgupta95) [inline app/controllers/users_controller.rb:42]: remove that instance variable, use local variable instead

File: app/controllers/application_controller.rb [hunk 8]
```diff
@@ -1,2 +1,43 @@
...
+  #   end
+  # end
+
+  # def authorize
+  #   render json: { message: 'You have to log in.' }, status: :unauthorized unless authorized_user
+  # end
+
+include JsonWebToken
+
+before_action :authenticate_request
```
Comment (neeteshgupta95) [inline app/controllers/application_controller.rb:34]: include that at the top of the file

**Review by neeteshgupta95 — CHANGES_REQUESTED** [review]

File: config/routes.rb (hunk 4 above)
Comment (Charchit507) [reply config/routes.rb:2]: sir only show, create, update and delete ke routes show krwane hai na

Comment (neeteshgupta95) [reply config/routes.rb]: ok

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
+class UsersController < ApplicationController
+  skip_before_action :authenticate_request, only: [:create]
+  before_action :set_user, only: [:show, :destroy]
+
+  def show
+    render json: user, status: :ok
```
Comment (neeteshgupta95) [inline app/controllers/users_controller.rb:6]: are you able to get the value of `user` here?

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+  def update
+    unless user.update(user_params)
+      render json: { errors: user.errors.full_messages },
+             status: :unprocessable_entity
+    end
+  end
+
+  def destroy
+    user.destroy
```
Comment (neeteshgupta95) [inline app/controllers/users_controller.rb:27]: are you able to get the value of user here?

File: config/routes.rb [hunk 11]
```diff
@@ -1,6 +1,4 @@
 Rails.application.routes.draw do
-  # Define your application routes per the DSL in https://guides.rubyonrails.org/routing.html
-
-  # Defines the root path route ("/")
-  # root "articles#index"
+  resources :users
```
Comment (neeteshgupta95) [inline config/routes.rb:2]: use only to those actions, which we are using

File: app/controllers/application_controller.rb [hunk 12]
```diff
@@ -1,2 +1,13 @@
...
+
+include JsonWebToken
+
+private
+  def authenticate_request
+      header = request.headers["Autherization"]
+      header = header.split(" ").last if header
+      decoded = jwt_decode(header)
+      @current_user = User.find(decoded[:user_id])
```
Comment (neeteshgupta95) [inline app/controllers/application_controller.rb:11]: do indentation

**Review by neeteshgupta95 — CHANGES_REQUESTED** [review]

**Review by neeteshgupta95 — APPROVED** [review]

**Pull request merged**